a9_fake_news_simulator/environment.py

```diff
@@ -157,9 +155,33 @@ def run_communication_protocol(self):
             sender_phonebook = [index for index in range(0, self.num_agents)
                                 if (sender_connectivity[index] != 0 and index != sender_index)]
             # randomly pick one of the possible receivers
-            receiver_index = sender_phonebook[random.randint(0, len(sender_phonebook)-1)]  # randint is inclusive
+            receiver_index = sender_phonebook[random.randint(0, len(sender_phonebook) - 1)]  # randint is inclusive
             receiver = self.agent_list[receiver_index]
             self.agent_conversation(sender, receiver)
+        if self.communication_protocol == "SYO":
+            # the name "learn new secrets" might be a bit misleading and should rather be "Spread your Opinion"
+            sender_index = self.choose_random_sender()
+            sender = self.agent_list[sender_index]
+            # choose receiver
+            # getting sender's connections
+            sender_connectivity = self.connectivity_matrix[sender_index, :]
+            # getting ids of possible receivers
+            sender_phonebook = [index for index in range(0, self.num_agents)
+                                if (sender_connectivity[index] != 0 and index != sender_index)]
+            # only select those contacts from which you know, that they don't know your opinion
+            valid_contacts = [contact for contact in sender_phonebook if
+                              self.agent_list[contact].opinion_base[sender_index] == 0]
```

Comment:
so right now I check whether the receiver never heard the opinion of the agent by checking against the opinion base entry 0.
However we could also check whether it is the opinion as the receiver has right now, because that might change over time. What do you guys think?

Reply:
The only way for an agent to know the opinion of others is to contact them, so checking against the value of the others' opinion could be the only right choice. However if you want to check for the SYO protocol, then anyone with a  different opinion should a valid/possible receiver, not only 0.
Of course all these decisions would be taken "to the best of an agent's knowledge", as usually happens with people, therefore one could contact an agreeing agent.